### PR TITLE
Add type testing, standardize test styles/coverage, simplify fallback schema

### LINKED ISSUE #63: test mocks and data - Improve maintainability and manageability of mock data
The current state of the tests are not consistent and while I love that the tests are there, I'm concerned that it will quickly become difficult to maintain. (especially at the recent pace of development).

This is a note to myself so that will focus on making the tests better and easier to add to for other users.

Any notes/comments welcome as I will probably not work on this for a few weeks.

### LINKED ISSUE #72: add type checking with vue-tsc
I've already started this but I want to log an issue so I can reference issue number.

### LINKED ISSUE #77: investigate: what happens if typegen is false (the default is true)
I'm struggling to think of why someone would want to use the module without the type generator but since it's an option I need to do more thinking about what happens if it's false. My guess is that we need to explain to the user how to name their Schema so that the rest of the wrapper handles their hand-made schema.

Probably a TODO for docs.

Issue created for me so I don't forget this one.

### PR DESCRIPTION
**Pull request by kheiner**
This branch adds vitest typecheck via `vue-tsc` for `.test-d.ts` files, fixes the errors that surfaced as a result, adds type contract tests, and cleans up some related docs inconsistencies.

**core module stuff**

- *typegenerator* `addTypeTemplate` was only called when `adminToken` was set and the Directus connection succeeded. If you ran `pnpm dev:prepare` without a `.env` file the `types/directus.d.ts` file was never written. necessary to ensure the ci pipeline works (since there won't be an .env in ci pipeline)
- *files composable* Removed a broken `import type { DirectusFile, DirectusSchema } from '#build/types/directus'`. That file uses `declare global` so named imports from it have never worked. Both types are already ambient globals. Changed the `data` field on `DirectusFileUpload` from `Partial<Record<keyof DirectusFile, string>>` to `Record<string, string>` because the old type collapsed to `{ id?: string }` under the fallback, rejecting valid metadata keys like `title` and `folder`. Added a `null` guard on `FormData.set()` and replaced a silent `undefined` return in `uploadDirectusFile` with a throw, making the return type a proper non-nullable `Promise<DirectusFile>`.

**playground stuff**

- *server* The server API routes were importing `defineEventHandler` and the module's server composables from `#imports`, which is the Nuxt client auto-import path. H3 utilities come from `h3` and the server composables need explicit source imports when type-checking under the root tsconfig rather than the Nitro-specific one. Added a larger than necessary comment so that it's obvious that the imports aren't required for them, only for our purpose.
- *shared demo schema* Added `playground/shared/demo.d.ts` with minimal stub interfaces for the collections the playground uses (`RolleyPost`, `RolleyPage`, `RolleyGlobals`) plus augmentations for `DirectusUser` and `DirectusSchema`.  The "Rolley" prefix matches the `types.prefix` setting in the playground config so the stubs are compatible with the generated types when a real Directus instance is connected. Also added another big note that outlines it's not necessary in their case.

**Docs**

- Fixed stale type names throughout the docs.  (plurals to singular that was added along time ago).  Also updated the `data` parameter type and return types for `uploadDirectusFile` and `uploadDirectusFiles` to match the changes from this PR.

**tests**

- *type-tests* Two new type contract test files. `test/auth.test-d.ts` asserts the public signatures of `useDirectusAuth` and `useDirectusUser`: return types, nullability, and that `UpdateMeInput` excludes `role` and `policies`. `test/files.test-d.ts` asserts that `uploadDirectusFile` resolves to a non-nullable `DirectusFile` and that `getDirectusFileUrl` accepts either a string ID or a `DirectusFile` object and returns `string`.
- Rewrote `test/known-issues-upstream.test-d.ts` to use a self-contained `IsolatedSchema` instead of relying on the global `DirectusSchema`. The previous version was inconsistent with pass/fail depending on the status of `.env` and reachability of the sandbox.

> [!NOTE]
> ~~My next big project is standardizing the tests and mock data so it's easier to maintain. Not sure when I'll get to that but this doesn't need to be pushed to next right away unless you want to do some live testing without a custom build.~~

**Scope expansion** — this branch absorbed `test/standardize-mock-and-tests` and `fix/types-disabled-edge-cases`, and includes follow-up work from the PR review thread.

**fallback type consolidation**

`fallback.d.ts` is now the single source of truth for fallback types. Previously a hardcoded `FALLBACK_TYPE_STRING` constant in `generate.ts` duplicated the same declarations. The module now reads `fallback.d.ts` from disk at startup and uses its content as the fallback template. `generate.ts` returns `null` instead of the fallback string when generation fails, and `module.ts` handles that `null` to decide which content to write. `addTypeTemplate` was also moved outside the `if (typesEnabled)` block — previously, if `options.types` was `false`, the template was never registered at all, which meant downstream code that imports `#build/types/directus` would silently break.

**fallback `DirectusFile` shape** (follow-up from [#78 review](https://github.com/rolleyio/nuxt-directus-sdk/pull/78#issuecomment-4346891162))

`fallback.d.ts` now extends `DirectusFile` with the string-scalar fields from the SDK via `Partial<Pick<SdkDirectusFile<object>, 'title' | 'description' | ...>>`. Without this, `Partial<Record<keyof DirectusFile, string>>` collapsed to `{ id?: string }` under the fallback, causing CI to reject valid metadata keys like `title`. The `data` parameter on `DirectusFileUpload` is back to `Partial<Record<keyof DirectusFile, string>>` per the review. A new SDK alignment type test in `test/files.test-d.ts` verifies that the listed fields still exist in the SDK's `DirectusFile` — `Pick` will error at compile time if the SDK renames one.

**test infrastructure**

Fixtures moved from `test/fixtures/directus-sdk/request/` to `test/fixtures/directus-sdk/` and a mock-builder pattern was introduced (`mockDirectusUpload().withVersion('latest')`). The `withVersion` API is versioned so that as Directus SDK responses evolve, older mock behaviour can be preserved by adding a new case above the current one rather than rewriting existing tests. A `test/README.md` documents how the suite is organised and how to add tests for new composables.

**auth composable test suite**

`test/auth.test.ts` was a `TODO` since issue #63. It is now fully implemented with the new mock-builder infrastructure, covering login, logout, `readMe`, `updateMe`, `createUser`, `inviteUser`, `acceptUserInvite`, `passwordRequest`, and `passwordReset`.

**rules fixture alignment**

`test/rules/fixtures/rules.json` and the surrounding test data were updated to match the Directus CMS sandbox template, keeping the fixture representative of a realistic schema rather than a synthetic one.

**docs**

Heading capitalisation normalised across the docs site. A new section on manually-created types was added to cover the case where typegen is disabled or unavailable, explaining the expected shape and the trade-offs.

---

- Fixes #72
- Fixes #63 
- Fixes #77

### CODE REVIEW

File: src/rules/loaders/json.ts [hunk 1]
```diff
@@ -341,12 +343,14 @@ export function loadRulesFromPayload<Schema>(
 
   // Build a map of policy ID -> PolicyConfig for role reference
   const policyById = new Map<string, PolicyConfig<Schema>>()
-  for (let i = 0; i < payload.policies.length; i++) {
-    const apiPolicy = payload.policies[i]
+  payload.policies.forEach((apiPolicy, i) => {
```
Comment (matt-rolley) [inline src/rules/loaders/json.ts]: Is there a preference here to use forEach?

File: src/rules/sync/format.ts [hunk 2]
```diff
@@ -162,7 +162,7 @@ function changePrefix(type: string): string {
  */
 function formatAddedFields(
   lines: string[],
-  item: Record<string, unknown>,
+  item: object,
```
Comment (matt-rolley) [inline src/rules/sync/format.ts:165]: Does object work better than Record? I've always thought it's better practice to use Record, but this could have changed?

File: src/rules/sync/push.ts [hunk 3]
```diff
@@ -89,17 +89,19 @@ export async function pushRules<Schema>(
 
     // Create new policies
     const policiesToCreate = diff.policies.filter(p => p.type === 'added')
-    for (let i = 0; i < policiesToCreate.length; i++) {
-      const change = policiesToCreate[i]
+    let policyCreateIdx = 0
+    for (const change of policiesToCreate) {
```
Comment (matt-rolley) [inline src/rules/sync/push.ts]: Not sure the benefit of this one over the previous? it just seems like we're introducing more variable's in the parent scope?

File: src/runtime/composables/files.ts [hunk 4]
```diff
@@ -1,17 +1,20 @@
-import type { DirectusFile, DirectusSchema } from '#build/types/directus'
 import type { DirectusFile as DirectusSdkFile, Query } from '@directus/sdk'
 import { uploadFiles } from '@directus/sdk'
 import { useDirectus, useDirectusUrl } from './directus'
 
 interface DirectusFileUpload {
   file: File
-  data?: Partial<Record<keyof DirectusFile, string>>
+  data?: Record<string, string>
```
Comment (matt-rolley) [inline src/runtime/composables/files.ts]: I think these lose types? If I added a company relation to a file, it'd be good for that to pull through? This way I could pass anything and fail it?

**Review by matt-rolley — COMMENTED** [review]
Just had a few comments, I'm not really against the changes I've commented, it'd just be good to know the reasons behind (for future use too)

File: src/rules/sync/format.ts (hunk 2 above)
Comment (kheiner) [reply src/rules/sync/format.ts:165]: You're right about the best practice; this was a blocker in getting the type check to pass and I (maybe incorrectly?) assumed that since it was only reading the object that it wouldn't make a functional difference.  I did notice that there's an improvement in the formatModifiedFields( function to enforce that local and remote should be of the same type <T extends object> 

Let me know if this makes sense or if you want me to try and build a more complex workaround.

File: src/rules/loaders/json.ts (hunk 1 above)
Comment (kheiner) [reply src/rules/loaders/json.ts]: This iterator wasn't prevalent in other parts of the codebase. There are a few other forEach loops so I modified. I don't think it introduces any actual improvements, except for readability/consistency across other parts of the code.

(the only `I++` strings in the code base were in push.ts and json.ts)
Screenshot 2026-04-29 at 08 36 45

(so moved it to forEach that exists)
Screenshot 2026-04-29 at 08 27 01

There's probably an improvement in refactoring further since we're already in a .map() but I went with the lazier consistency fix.

File: src/rules/sync/push.ts (hunk 3 above)
Comment (kheiner) [reply src/rules/sync/push.ts]: Good catch, it was a stylistic choice that I made (driven by the same as above in rules/loaders/json.ts). I think there's a for (const [idx, change] of ... policiesTo<Delete/Create...>.entries() 

that addresses both your concern and my desire for stylistic consistency.

Comment (matt-rolley) [reply src/rules/sync/push.ts]: Yeah this would be better :) thank you!

File: src/rules/sync/format.ts (hunk 2 above)
Comment (matt-rolley) [reply src/rules/sync/format.ts:165]: I've always just found using object on it's own becomes an issue further down the line, like in this example:

Screenshot 2026-04-29 at 17 06 36

But if that's not the case, happy to use it

File: src/rules/loaders/json.ts (hunk 1 above)
Comment (matt-rolley) [reply src/rules/loaders/json.ts]: Happy to use forEach here, I remember seeing a while go that forEach was slower than for loops, but ultimately we've probably got better ways to increase performance than this small increase in speed. So for readability, it's probably worth doing this.

I think as long as we try to stick to one way of doing it, forEach/map/reduce vs for then that works for me :)

File: src/runtime/composables/files.ts (hunk 4 above)
Comment (kheiner) [reply src/runtime/composables/files.ts]: Oops. This was remnants of a workaround before I dug deeper into the fallback.d.ts workaround. You're right that this is an issue and should be reverted.

File: src/rules/sync/format.ts (hunk 2 above)
Comment (kheiner) [reply src/rules/sync/format.ts:165]: understood, I don't think that situation applies here. I can spend time thinking about a more complex workaround if you think it will apply here.

Please let me know what path you want to take. For now I'm going to assume it's safe to keep as an object.

**Comment by kheiner** [comment]
I think having the multiple branches with separate PRs was creating more confusion on my end than it was serving the separated concerns. I merged those branches into this for one huge PR.

The one thing that I hope doesn't get lost here is the comment https://github.com/rolleyio/nuxt-directus-sdk/pull/78/changes/BASE..1bbed3127218cbd665461e06b19c075f6bd2bc6e#r3162499634 regarding object type vs Record<...>.

This also ensures that it's easier to action any feedback you have, but I am truly sorry for the massive amount of changes I'm asking you to review. (now that I'm looking at it more closely, the big line changes are just mock data)

As far as finishing touches, I don't know if this should be 6.0 - it seems that 5.1 is the right answer -- I'd like to reserve major version for non-backward compatibility and I think this still has backwards compatibility with what's in 5.

> [!NOTE]
> I will be doing a full second review in GitHub to provide a "third" set of eyes here and get ahead of any questions you might have about stylistic changes.

File: test/rules/example-rules.test.ts [hunk 5]
```diff
@@ -10,104 +10,113 @@ import {
...
 describe('diff functionality', () => {
   it('shows diff between remote and extended rules', () => {
-    // Serialize our extended rules to the Directus API format
     const localPayload = serializeToDirectusApi(rules)
-
-    // Compare with the original remote payload
     const diff = compareRulesPayloads(localPayload, payload)
-
-    // Print the diff
+    // TODO: Does this need to be printed to the console? ask @matt-rolley
```
Comment (kheiner) [inline test/rules/example-rules.test.ts]: Does the formatDiff() need to be in the console log? it's programatically checked so visualizing the diff doesn't feel necessary to me. It's not a huge deal to me but just wondering how you're using that part.

File: test/rules/integration.test.ts
Comment (kheiner) [inline test/rules/integration.test.ts:1]: I have a stashed version of this file that contemplates matching it more closely with the CMS sandbox, but the tests here provided better coverage. It inspired me to try and get the cms template updated to have the ability to put comments under blogs so we can retain the template and keep the coverage.

https://github.com/directus-labs/directus-templates/issues/22

File: docs/api/composables/file.md [hunk 6]
```diff
@@ -9,10 +9,10 @@ outline: deep
 Upload a single file to Directus.
 
 **Parameters:**
-- `file: { file: File, data?: Partial<DirectusFiles> }` - File and metadata
+- `file: { file: File, data?: Record<string, string> }` - File and metadata
```
Comment (kheiner) [inline docs/api/composables/file.md]: TODO: Doublecheck

File: src/runtime/types/fallback.d.ts
Comment (kheiner) [inline src/runtime/types/fallback.d.ts:1]: REVISIT: What code gets simplified for type checking if we...

```ts
interface DirectusSchema {
  directus_files: DirectusFile
  directus_user: DirectusUser
}
```

My thought is that it may prevent us from needing to have the less clear 'extends' on this - as the extending is done by the SDK. I did this because it was required to get the typecheck to pass but I don't remember testing the case where the SDK could possibly do the extend for me.

This shouldn't be a blocker in pushing the code as this definitely works, it could be clearer though. so I'll investigate further.

File: test/fixtures/directus-sdk/read-fields.data.ts
Comment (kheiner) [inline test/fixtures/directus-sdk/read-fields.data.ts:1]: Discussion here: https://github.com/rolleyio/nuxt-directus-sdk/issues/79

File: test/files.test-d.ts
Comment (kheiner) [inline test/files.test-d.ts:1]: Related: https://github.com/rolleyio/nuxt-directus-sdk/pull/78/changes#r3168853499

File: test/files.test.ts [hunk 7]
```diff
@@ -1,154 +1,143 @@
-import type { DirectusFile } from '@directus/sdk'
 import { beforeEach, describe, expect, it, vi } from 'vitest'
-import { readFiles } from './fixtures/directus-sdk/request/read-files.data'
+import { readFiles } from './fixtures/directus-sdk/read-files.data'
+import { mockDirectusUpload, requestMock, state, uploadFilesMock } from './fixtures/directus-sdk/upload-files.mock'
+import { getDirectusFileUrl, uploadDirectusFile, uploadDirectusFiles } from '../src/runtime/composables/files'
 
 const MOCK_BASE_URL = 'https://cms.example.com'
```
Comment (kheiner) [inline test/files.test.ts:6]: Check for consistency of mock urls

File: test/rules/example-rules.test.ts (hunk 5 above)
Comment (kheiner) [reply test/rules/example-rules.test.ts]: I removed the console log in my latest commit batch, if there's a purpose here let me know and it can easily be restored.

File: src/runtime/types/fallback.d.ts
Comment (kheiner) [reply src/runtime/types/fallback.d.ts:1]: Investigatino complete, it didn't help as much as I thought it would, but it did help. so added.

**Comment by kheiner** [comment]
I made changes and left the comments open that I thought you might have an opinion on. Re-requesting review.

P.S. I think the version control should keep this on v5, not v6 but I don't remember if I communicated that here or somewhere else.

**Comment by matt-rolley** [comment]
Hey, yeah lots of changes here, I think I've gone through most of them before but will take another look. I'm not sure if it'll be possible to do it as 5.1, some of the changes here reference breaking changes, like useSessionDirectus, so if you wanted to do a PR for 5.1 with those taken out then a seperate one to update it after, that'd be good. Otherwise, I think we just try to get 6.0 live asap (once we're happy with testing) then hopefully not too many people used V5?

File: src/rules/sync/format.ts (hunk 2 above)
Comment (matt-rolley) [reply src/rules/sync/format.ts:165]: Yeah agreed, looking at it again the `<T extends object>` is doing the useful work here (pinning local and remote to the same shape) and switching the public param to `Record<string, unknown>` would just shift the `as` cast somewhere else. Happy to leave as-is, resolving.

File: test/rules/example-rules.test.ts (hunk 5 above)
Comment (matt-rolley) [reply test/rules/example-rules.test.ts]: No need to restore it — it was just noise in passing test output. Removal is the right call. Resolving.

File: test/rules/integration.test.ts
Comment (matt-rolley) [reply test/rules/integration.test.ts:1]: Makes sense to keep the current fixture for the coverage win. Nice catch on the upstream template too — happy to revisit once [directus-labs/directus-templates#22](https://github.com/directus-labs/directus-templates/issues/22) lands. Resolving.

File: test/fixtures/directus-sdk/read-fields.data.ts
Comment (matt-rolley) [reply test/fixtures/directus-sdk/read-fields.data.ts:1]: I'd keep them whole for now — the "real data catches things synthetic data hides" benefit is worth the size, and token cost only matters when an agent is reading the file. The AGENTS.md note you suggested in #79 sounds like the right fix if/when it actually becomes a friction point. Resolving this thread, let's keep #79 open for the broader conversation.

**Comment by matt-rolley** [comment]
Thanks for the thorough self-review pass and for absorbing the other branches into this one, easier to reason about as a single change. Most of this lands well; four open threads resolved above.

On version: keeping this on 6.0 rather than 5.1. `useSessionDirectus` and the related renames already on `next` are breaking, so this should ride `6.0.0-beta.3`. I'll create a new prerelease once this is merged.

One small thing worth calling out in the changelog when we cut the beta: `uploadDirectusFile` now throws instead of returning `undefined` on a failed upload, a minor behaviour change for anyone catching the old silent path.

Let me know what you think and if you're happy to merge?

**Comment by kheiner** [comment]
> On version: keeping this on 6.0 rather than 5.1. `useSessionDirectus` and the related renames already on `next` are breaking, so this should ride `6.0.0-beta.3`. I'll create a new prerelease once this is merged.

I completely forgot about that; good call on 6.0


> One small thing worth calling out in the changelog when we cut the beta: `uploadDirectusFile` now throws instead of returning `undefined` on a failed upload, a minor behaviour change for anyone catching the old silent path.

Since everything is pushed any amend I do is forced. Should I do an empty commit to get changelogen to add the entry?

> Let me know what you think and if you're happy to merge?

Yes. I think there will be small bugs that might get through because it's such a big change but I feel as if I've thoroughly tested against the major cases.
